File: day3/day3.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;


long toboggan (const vector<string> hillMap, int dx, int dy)
{
  int trees=0, x=dx, y=dy;
  int height = hillMap.size();
  int width = hillMap[0].length();

  while(y<height)
  {
  //  cout << " x=" << x << " y= " << y << " " << hillMap[y][x % width] << endl;
    if (hillMap[y][x % width]=='#') trees++;
    x+=dx;
    y+=dy;
  }

  return trees;
}

int main() 
{
    ifstream inputFile;
    vector<string> hillMap;

    inputFile.open("input.txt", ios::in);
    while (!inputFile.eof()) 
    {
      string line;
      getline(inputFile,line);
      hillMap.push_back(line);
    }

    long trees = toboggan(hillMap,3,1);
    cout << "Part 1: Trees = " << trees << endl;

    trees = trees * toboggan(hillMap,1,1)
                  * toboggan(hillMap,5,1)
                  * toboggan(hillMap,7,1)
                  * toboggan(hillMap,1,2);
    cout << "Part 2: Tree Product = " << trees << endl;

    return 0;
}

File: day3/day3.py
hillfile =  open('day3/input.txt','r')
hill=[]
for line in hillfile:
  hill.append(line.strip())

height = len(hill)
width = len(hill[1])
print("Map size = ({},{})".format(width,height))

slopes = [[1,1],[3,1],[5,1],[7,1],[1,2]]

prod = 1
for slope in slopes:
    dx = slope[0]
    dy = slope[1]
    x = 0
    y = 0
    trees=0
    for y in range(dy, height, dy):
      x+=dx
      if hill[y][x % width]=="#": trees+=1
    print("Slope ({},{}) => {} trees".format(dx,dy,trees))
    prod *=trees
    
print("Tree product for all slope = {}".format(prod))